sdk/go/src/sawtooth_sdk/messaging/stream.go

```diff
@@ -105,107 +164,132 @@ func (self *Stream) Respond(t validator_pb2.Message_MessageType, c []byte, corrI
 	}
 
 	self.outgoing <- request
+
+	return request.Response
 }
 
+// Setup the sockets and channels needed for communicating.
 func (self *Stream) start() {
 	socket, err := self.context.NewSocket(zmq.DEALER)
 	if err != nil {
-		fmt.Sprint("Failed to create ZMQ socket: ", err)
+		logger.Error("Failed to create ZMQ socket: ", err)
 		return
 	}
 
 	socket.SetIdentity(self.identity)
-	fmt.Println("Socket Identity set to", self.identity)
+	logger.Debug("Socket Identity set to ", self.identity)
 
-	fmt.Printf("Connecting to %v...", self.url)
+	logger.Info("Connecting to ", self.url)
 	err = socket.Connect(self.url)
 	if err != nil {
 		socket.Close()
-		fmt.Sprint("Failed to connect to ", self.url, ": ", err)
+		logger.Error("Failed to connect to ", self.url, ": ", err)
 		return
 	}
-	fmt.Println("done")
 	self.socket = socket
 
 	reactor := zmq.NewReactor()
 	reactor.AddSocket(self.socket, zmq.POLLIN, self.receiver)
 	reactor.AddChannel(self.outgoing, SENDS_PER_LOOP, self.sender)
 	err = reactor.Run(POLLING_TIMEOUT)
-	fmt.Println("Reactor exited:", err)
+	logger.Error("Reactor exited: ", err)
 }
 
-// Send messages from the outgoing channel in a loop
+// Handle a single message sendRequest
 func (self *Stream) sender(i interface{}) error {
+	// Handle a panic while sending without crashing. For example, if the
+	// channel passed in with the sendRequest is closed.
+	defer func() {
+		if r := recover(); r != nil {
+			logger.Warn("Panic while sending: ", r)
+		}
+	}()
+
+	// Validate this is a sendRequest.
 	req, ok := i.(*sendRequest)
 	if !ok {
-		fmt.Println("Received unexpected type from channel!")
+		logger.Warn("Received unexpected type from channel!")
 		return nil
 	}
 
+	// Deserialize
 	msgData, err := proto.Marshal(req.Msg)
 
 	if err != nil {
-		// TODO: Handle a panic in the event the channel is already closed
 		req.Response <- &Response{
 			Msg: nil,
 			Err: &SendMsgError{fmt.Sprint("Failed to marshal:", err)},
 		}
 		close(req.Response)
+		return nil
 	}
 
+	// Send the message
 	err = sendBytes(self.socket, msgData)
 	if err != nil {
-		// TODO: Handle a panic in the event the channel is already closed
 		req.Response <- &Response{
 			Msg: nil,
 			Err: &SendMsgError{fmt.Sprint("Failed to send message:", err)},
 		}
 		close(req.Response)
+		return nil
 	}
 
 	// Store the channel so the response can be sent on it
-	if !req.IsResponse {
+	if req.IsResponse {
+		req.Response <- &Response{
+			Msg: nil,
+			Err: nil,
+		}
+		close(req.Response)
+	} else {
 		self.responses[req.Msg.CorrelationId] = req.Response
 	}
 	return nil
 }
 
-// Receive messages in a loop and route
+// Receive a single message and route
 func (self *Stream) receiver(socketState zmq.State) error {
-	// 1. Receive message
+	// Handle a panic while received without crashing. For example, if the
```

Comment:
Usage? 
Change : Handle a panic while received without
to
 Handle a panic while receiving without